### ISSUE TITLE
Possibility of removing packages from Hackage

### ISSUE THREAD

**Issue by jonkri** [haskell/hackage-server#112]
I am a little unhappy about that fact that some deprecated packages that I wrote way back are shown on the new Hackage when I search for "pontarius":

http://hackage.haskell.org/packages/search?terms=pontarius

The packages also show up on http://hackage.haskell.org/packages/ (not even being marked as deprecated).

I'd really like for authors to be able to remove packages from Hackage completely.

**Comment by sol** [comment]
I agree that tarballs should not be removed.  But other than that I think I as an author of a package should have the right and means to unpublish a package.

**Comment by hvr** [comment]
@sol it appears only under `http://hackage.haskell.org/user/<username>` if you're in its maintainer-group... so you'd simply have to leave the maintainer group of that package

@sol also, what about http://hackage.haskell.org/packages/deprecated ? would you want that index to be destroyed as well?

**Comment by sol** [comment]
@hvr Good point regarding the maintainer-group.

Regarding the list of "All deprecated packages", I see how this can be *interesting*, but I'm still puzzled in which scenarios this is *useful*.

**Comment by andreasabel** [comment]
If you are moving to a new house, you can keep everything for a while, but after some years, you have to start thinking how to sort out some things that just sit there collecting dust.

I think for hackage the time has come where we have to think about procedures for spring cleaning.
Packages that haven't been buildable for years and are abandoned by the maintainers and forgotten by the world should be moved to an _attic_ space rather that continue to sit in the main hackage directory.  Why?  Is disk space a problem?  Well, no, most can still afford 500MB for the package index (as of Nov 2018), but:

* Old, abandoned packages are squatting nice package names.
* The Haskell developer looking for a library for a certain purpose has to search through a pile of trash to find the gem.
* Having no quality control at all on the central package server does not shine a good light on the Haskell community.

I think packages that have not compiled for many years should lose their right to remain on hackage. They can be moved to some attic space where they can remain for archeological purposes or wait for their resurrection.  
Weeding out the dead packages would be a first step to a saner hackage (one could think of further quality ensurance mechanisms later, like reviewing or ranking by reverse dependencies...).

[Maybe I should open a new issue for my proposal?!]

**Comment by gbaz** [comment]
I think the need for a clean presentation should be covered by https://github.com/haskell/hackage-server/issues/680

I have that assigned to me but if someone wants to take it over, it would be welcome.

In short -- we could establish standard "filter" criteria that would screen out packages that hadn't been uploaded in however long, or that didn't build with newer ghcs, etc. This can all be done by augmenting our current table display with some additional javascript.

**Comment by kolmodin** [comment]
> If you are moving to a new house, you can keep everything for a while, but after some years, you have to start thinking how to sort out some things that just sit there collecting dust.
> 
> I think for hackage the time has come where we have to think about procedures for spring cleaning.
> Packages that haven't been buildable for years and are abandoned by the maintainers and forgotten by the world should be moved to an _attic_ space rather that continue to sit in the main hackage directory. Why? Is disk space a problem? Well, no, most can still afford 500MB for the package index (as of Nov 2018), but:
> 
> * Old, abandoned packages are squatting nice package names.
> * The Haskell developer looking for a library for a certain purpose has to search through a pile of trash to find the gem.
> * Having no quality control at all on the central package server does not shine a good light on the Haskell community.
> 
> I think packages that have not compiled for many years should lose their right to remain on hackage. They can be moved to some attic space where they can remain for archeological purposes or wait for their resurrection.
> Weeding out the dead packages would be a first step to a saner hackage (one could think of further quality ensurance mechanisms later, like reviewing or ranking by reverse dependencies...).
> 
> [Maybe I should open a new issue for my proposal?!]

I disagree that we should remove packages. 

- Most package names are not taken, just pick a new one. When packages are forked people come up with new names, it's not an issue. What is a nice package name anyway? Are attoparsec and criterion bad names? Just name your package whatever you want, if it's a good package it'll be used :)
- If discovery of good packages is hard then that's a problem thats bigger than having to look at some old packages. Search should be improved rather than pruning the package database. That would solve visibility of good packages, demote unpopular packages (broken, old or just plain bad).
- A package cannot be resurrected, as you suggest, if its package name has been taken.

It's seems that at the core of this issue is the problem of finding good packages, not that we also have bad packages. IMO we should spend time on promoting good packages (with some package ranking) rather than spending time on completely removing a small set of packages and keep the poor search we have.
Spring cleaning is after all just a ranking where you highlight the useful stuff and demote things you think will be less useful in the future.
